File: neet_code_all_cpp/array_hash/concatenation_of_array.cpp
#include <iostream>
#include <vector>

//AC
class Solution {
    public:
        std::vector<int> getConcatenation(std::vector<int>& nums) {
            auto n = nums.size();
            std::vector<int> result(2*n);

            for(int i = 0; i < n; ++i) {
                result[i] = nums[i];
                result[i + n] = nums[i];
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<int> getConcatenation(std::vector<int>& nums) {
            std::vector<int> ans;
            auto len = nums.size();
            for(int i = 0; i < 2 *len; ++i) {
                ans.push_back(nums[i % len]);
            }

            return ans;
        }
};

void show_result(std::vector<int>& result) {
    for(auto& v: result) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    auto case_1 = std::vector{1, 2, 1};
    auto case_2 = std::vector{1, 3, 2, 1};

    Solution s_1;
    auto res_1 = s_1.getConcatenation(case_1);
    auto res_2 = s_1.getConcatenation(case_2);
    show_result(res_1);
    show_result(res_2);
}

File: neet_code_all/src/array_hash/concatenation_of_array.rs
struct Solution {}
impl Solution {
    // AC
    pub fn get_concatenation(nums: Vec<i32>) -> Vec<i32> {
        [nums.clone(), nums].concat()
    }

    // AC(こっちのほうが速い)
    pub fn get_concatenation_2(nums: Vec<i32>) -> Vec<i32> {
        let mut result = nums.clone();
        for v in nums {
            result.push(v);
        }

        result
    }
}

// 模範解答:
struct SolutionAns {}
impl SolutionAns {
    pub fn get_concatenation(nums: Vec<i32>) -> Vec<i32> {
        let mut result = vec![0; 2 * nums.len()];
        for i in 0..nums.len() {
            result[i] = nums[i];
            result[i + nums.len()] = nums[i];
        }

        result
    }
}

fn main() {
    let case_1 = vec![1, 2, 1];
    let case_2 = vec![1, 3, 2, 1];

    println!("case_1: {:?}", Solution::get_concatenation(case_1.clone()));
    println!("case_2: {:?}", Solution::get_concatenation(case_2.clone()));

    println!(
        "case_1: {:?}",
        SolutionAns::get_concatenation(case_1.clone())
    );
    println!(
        "case_2: {:?}",
        SolutionAns::get_concatenation(case_2.clone())
    );
}
